sourcing/domain/service.go

```diff
@@ -0,0 +1,318 @@
+		// throws ErrNeedMoreDetailsSourceCannotBeDetected if information on the cart (or delivery is missing)
+		AllocateItems(ctx context.Context, decoratedCart *decorator.DecoratedCart) (ItemAllocations, error)
+
+		// GetAvailableSources returns possible Sources for the product and the desired delivery.
+		// Optional the existing cart can be passed so that existing items in the cart can be evaluated also (e.g. deduct stock)
+		// e.g. use this before a product should be placed in the cart to know if and from where an item can be sourced
+		// throws ErrNeedMoreDetailsSourceCannotBeDetected
+		// throws ErrNoSourceAvailable if no source is available for the product and the given delivery
+		GetAvailableSources(ctx context.Context, product domain.BasicProduct, deliveryInfo *cartDomain.DeliveryInfo, decoratedCart *decorator.DecoratedCart) (AvailableSources, error)
+	}
+
+	// ItemID string alias
+	ItemID string
+
+	// ItemAllocations represents the allocated Qtys per itemId
+	ItemAllocations map[ItemID]ItemAllocation
+
+	//ItemAllocation info
+	ItemAllocation struct {
+		AllocatedQtys AllocatedQtys
+		Error         error
+	}
+
+	// AllocatedQtys represents the allocated Qty per source
+	AllocatedQtys map[Source]int
+
+	// Source descriptor for a single location
+	Source struct {
+		// LocationCode identifies the warehouse or stock location
+		LocationCode string
+		// ExternalLocationCode identifies the source location in an external system
+		ExternalLocationCode string
+	}
+
+	// AvailableSources is the result value object containing the available Qty per Source
+	AvailableSources map[Source]int
+
+	//DefaultSourcingService - an example implementation
+	DefaultSourcingService struct {
+		availableSourcesProvider AvailableSourcesProvider
+		stockProvider            StockProvider
+		logger                   flamingo.Logger
+	}
+
+	// AvailableSourcesProvider interface for DefaultSourcingService
+	AvailableSourcesProvider interface {
+		GetPossibleSources(ctx context.Context, product domain.BasicProduct, deliveryInfo *cartDomain.DeliveryInfo) ([]Source, error)
+	}
+
+	// StockProvider interface for DefaultSourcingService
+	StockProvider interface {
+		GetStock(ctx context.Context, product domain.BasicProduct, source Source) (int, error)
+	}
+)
+
+var (
+	_ SourcingService = new(DefaultSourcingService)
+
+	// ErrInsufficientSourceQty - use to indicate that the requested qty exceeds the available qty
+	ErrInsufficientSourceQty = errors.New("Available Source Qty insufficient")
+
+	// ErrNoSourceAvailable - use to indicate that no source for item is available at all
+	ErrNoSourceAvailable = errors.New("No Available Source Qty")
+
+	// ErrNeedMoreDetailsSourceCannotBeDetected - use to indicate that informations are missing to determine a source
+	ErrNeedMoreDetailsSourceCannotBeDetected = errors.New("Source cannot be detected")
+)
+
+//Inject the dependencies
+func (d *DefaultSourcingService) Inject(
+	logger flamingo.Logger,
+	dep *struct {
```

Comment:
why are the dependencies optional here? The service will not work without them.

Reply:
The dependencies are optional as we wanted to have only error reporting, but not startup failures if the required services are not configured.